packages/typegpu/src/std/numeric.ts

```diff
@@ -29,13 +30,50 @@ export function sub<T extends vBase>(lhs: T, rhs: T): T {
   return VectorOps.sub[lhs.kind](lhs, rhs);
 }
 
-export function mul<T extends vBase | AnyMatInstance>(s: number | T, v: T): T {
+export function mul<T extends AnyMatInstance, TVec extends vBaseForMat<T>>(
+  s: T,
+  v: TVec,
+): TVec;
+export function mul<T extends AnyMatInstance, TVec extends vBaseForMat<T>>(
+  s: TVec,
+  v: T,
+): TVec;
+export function mul<T extends vBase | AnyMatInstance>(s: number | T, v: T): T;
+export function mul(
+  s: vBase | AnyMatInstance | number,
+  v: vBase | AnyMatInstance,
+): vBase | AnyMatInstance {
   if (inGPUMode()) {
-    return `(${s} * ${v})` as unknown as T;
+    return `(${s} * ${v})` as unknown as vBase | AnyMatInstance;
   }
   if (typeof s === 'number') {
+    // Scalar * Vector/Matrix case
     return VectorOps.mulSxV[v.kind](s, v);
   }
+  if (
+    typeof s === 'object' &&
+    typeof v === 'object' &&
+    'kind' in s &&
+    'kind' in v
+  ) {
+    const sIsVector = !s.kind.includes('mat');
+    const vIsVector = !v.kind.includes('mat');
```

Comment:
Tiny nit, covers all the cases and might be more performant.
```suggestion
    const sIsVector = !s.kind.startsWith('mat');
    const vIsVector = !v.kind.startsWith('mat');
```